### ISSUE TITLE
Issues connecting with DocumentDB Emulator using Node SDK

### ISSUE THREAD

**Issue by jagratimodi** [Azure/azure-documentdb-node#157]
We're trying to connect to DocumentDB emulator using the Node SDK (version 1.10.0), we encountered few issues:

1. First issue was with the self-signed certificate. For some reason the HTTP request was declined because of this. We searched on how to work with self-signed certificate, one way is to make use of environment variable “NODE_TLS_REJECT_UNAUTHORIZED” and setting this variable to “0”. But a safer way to solve this issue is by adding the following line of code in the SDK(request.js file):
   requestOptions.rejectUnauthorized = false;
After adding this line, we no longer receive the error with self-signed certificate issue but now we get an error saying the service is not running.
   
2. Second issue, Node’s HTTPS library is pretty choosy about the port. Unless explicitly overridden, HTTPS library expects port to be 443. So to make things work, we added the following line of code in the SDK to set default port of emulator service.(request.js file after line 172):
     requestOptions.port = 8081;

The following values can be set in SDK and the “EnableEndpointDiscovery” to false in the ConnectionPolicy while locally connecting to DocumentDB Emulator using Node SDK:
  requestOptions.rejectUnauthorized = false; //For self-signed certificates
  requestOptions.port = 8081; //Setting default port for DocumentDB Emulator

**Comment by rnagpal** [comment]
@jagratimodi 
1. We fixed the self-signed certificate issue with the 1.10.1 release of the SDK, so you should be good there.

Can you please elaborate on:
>>but now we get an error saying the service is not running.
Are you seeing Service Unavailable exception? Can you send us the complete stack trace?

2. You shouldn't have to do change the port explicitly. We had a port issue in our first release of the emulator but we fixed it soon. Please make sure you are using the latest emulator from https://aka.ms/documentdb-emulator and using 8081 in the Url as: https://localhost:8081.
You also don't have to set the EnableEndpointDiscovery to false explicitly(this was the same issue in our first emulator release).

So in summary, please use 1.10.1 version of the Node.js SDK, with no workarounds you mentioned above when using the latest emulator and you should be able to connect. Let us know if that's not the case.

**Comment by moderakh** [comment]
I am closing this issue as this is fixed in the latest release of Emulator and NodeJS SDK. If you think still that's an issue feel free to re-open.

**Comment by MatejSkrbis** [comment]
I updated my npm packages from 1.10.2 to 1.11.0 and it looks like the issue is back or at least it's similar.
The error returned is:
```
{
    code: "DEPTH_ZERO_SELF_SIGNED_CERT",
    error: "self signed certificate"
}
```

**Comment by rnagpal** [comment]
@MatejSkrbis With the latest release of emulator, the hostname returned from the server cannot be always "localhost" so we had to make this change. Now you can set DisableSSLVerification option on ConnectionPolicy object to true when connecting to emulator. Let me know if that resolves this issue.

**Comment by MatejSkrbis** [comment]
I added the DisableSSLVerification option on ConnectionPolicy object, but now it returns different error:

```
{
    "code":"BadRequest","message":"Invalid API version. Ensure a valid x-ms-version header value is passed.\r\nActivityId: dfce9f06-da1e-4d08-9184-8d401cb1263f"
}
```

```
const { DocumentClient, DocumentBase } = require('documentdb');
...
const connectionPolicy = new DocumentBase.ConnectionPolicy();
connectionPolicy.DisableSSLVerification = true;

const client = new DocumentClient(config.host, { masterKey: config.authKey }, connectionPolicy);
```

**Comment by rnagpal** [comment]
@MatejSkrbis Yes. You seemed to have an older version of the emulator but the latest client and hence this error was expected.

**Comment by GSchroder** [comment]
I am getting this problem still, I am running v1.11.0 of the DocumentDB npm package, on Node.js v7.8.0, and DocumentDBEmulator v1.11.136.2.

**Comment by rnagpal** [comment]
@GSchroder Not sure which error you are talking about: SSL verification one or Invalid API version.
If it's former:
Did you set connectionPolicy.DisableSSLVerification = true; in your client?
You now need to explicitly disable SSL verification when connecting to DocumentDB emulator as above.

**Comment by rnagpal** [comment]
FYI to everyone of this thread. The issue that we fixed in 1.10.1 to allow connections to emulator automatically doesn't work with the latest releases of emulator as the design has changed, so now you need to explicitly set DisableSSLVerification to true like:
connectionPolicy.DisableSSLVerification = true; in your client to avoid getting Self signed certificate error.

**Comment by GSchroder** [comment]
Apologies for not being clear, it was the SSL verification issue. Thanks for the info, I didn't realise the design had changed to manually disabling it now.
